### PR TITLE
Dev adam graph conf

### PR DESCRIPTION
**Pull request by MARD1NO**
给Adam optimizer增加graph所需的conf

增加adam，sgd对应的单测脚本

### CODE REVIEW

File: python/oneflow/nn/optimizer/adam.py [hunk 1]
```diff
@@ -138,3 +139,38 @@ def step(self, closure: Callable = None):
...
+                base_scale, scale
+            )
+
+            train_conf.set_loss_scale_factor(scale)
+            optimizer_conf.set_base_learning_rate(lr)
+
+            optimizer_conf.mutable_adam_conf().set_beta1(beta1)
+            optimizer_conf.mutable_adam_conf().set_beta2(beta2)
+            optimizer_conf.mutable_adam_conf().set_epsilon(epsilon)
+            optimizer_conf.mutable_adam_conf().set_do_bias_correction(
```
Comment (MARD1NO) [inline python/oneflow/nn/optimizer/adam.py:168]: 这里设置为False的原因是，对应eager调用的op在 https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/user/kernels/model_update_kernel_util.h#L81-L89

可以看到这里是没有做bias_correction的

Comment (Ldpe2G) [reply python/oneflow/nn/optimizer/adam.py:168]: https://github.com/Oneflow-Inc/oneflow/blob/95337ebc20fa93c28b67985c00aa24b98eb46d78/oneflow/core/job_rewriter/adam_optm.cpp#L134

bias_correction 的逻辑是在这里处理的，不过旧的 lazy 接口默认也是 false

https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/ops/optimizer.py#L1344

File: python/oneflow/test/graph/test_graph_optimizer.py [hunk 2]
```diff
@@ -72,6 +72,54 @@ def build(self, x):
...
+                out.backward()
+                return out
+
+        g = CustomGraph0()
+        x = flow.Tensor(4, 10)
+        flow.nn.init.uniform_(x, a=-1.0, b=1.0)
+        z = g._compile(x)
+        print("repr(g): \n", repr(g))
+        print("g.config.proto: \n", g.config.proto)
+        print("graph proto: \n", g._graph_proto)
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_optimizer.py:121]: 这个optimizer test之前做的不严谨，因为当时功能不全，指验证了conf的生成。可以改成训练nn.Linear Module。

参考：graph测试下的test_graph_linear_train.py

Comment (Ldpe2G) [reply python/oneflow/test/graph/test_graph_optimizer.py:121]: 是的，而且应该是参考 lazy 和 eager  optimzer 的测试
https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/compatible/single_client/test/ops/test_optimizers.py#L929

https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/test/modules/test_optim_adam.py

File: python/oneflow/test/graph/test_graph_adam_optim.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                    "scale": scale,
+                }
+            ]
+        )
+
+    class CustomAdamGraph(flow.nn.Graph):
+        def __init__(self, grad):
+            super().__init__()
+            self.m = simp_module
+            self.grad_tensor = flow.Tensor(
```
Comment (MARD1NO) [inline python/oneflow/test/graph/test_graph_adam_optim.py:57]: 这里遇到的一个问题是，想对齐原来的测试 https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/compatible/single_client/test/ops/test_optimizers.py#L160-L188


他每次都会给global_function传入一个tensor，和里面的Variable相乘，然后再求loss，求梯度

由于graph在一开始就需要构造好module，所以我把这个输入变量放到module的forward函数（原因是这个输入变量是取自random_grad_seq[i]，在68行内的for循环里取到。我不可能把这个输入变量放到module的init初始化，如果这样module的构造，graph的构造都要放到for循环里了）

定义好图后，我想是在graph的init阶段，把grad转换为tensor，并给到module的forward函数。

但是这么做会导致 
```
oneflow._oneflow_internal.exception.UnimplementedException: stack_frame {
  location: "/home/zzk/oneflow/oneflow/core/framework/op_interpreter/lazy_op_interpreter.cpp:325"
  function: "ApplyImpl"
  error_msg: " Sorry! nn.Graph does NOT support free eager tensor which NOT captured by nn.Graph. Please using nn.Graph.build() args or nn.Module.__init__() to handle all eager tensor."
```

File: python/oneflow/test/graph/test_graph_adam_optim.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    simp_module = CustomModule()
+    simp_module.to("cuda")
+    simp_module.train()
+
+    adam0 = flow.optim.Adam(
+            [
+                {
+                    "params": simp_module.parameters(),
+                    "lr": learning_rate,
```
Comment (MARD1NO) [inline python/oneflow/test/graph/test_graph_adam_optim.py:44]: 我这里有个关于adam设置学习率的问题，我按照xiaoyu给的sgd示例，给adam增加generate_conf_for_graph这个函数。
然后我跑到`oneflow/user/kernels/model_update_kernel_util.h` AdamUpdateFunctor这里打log
    *v = next_v;
    *model =
        model_val - learning_rate * (next_m / (sqrt(next_v) + epsilon) + weight_decay * model_val);
    ...
    printf("==||| learning rate: %f \n", learning_rate); // Here!!!!!
然后我发现我在外面设置的是learning_rate = 1，这里打印结果是31.622810
设置learing_rate = 0.0001，打印结果是0.003162
就是这里的问题引发了计算对不齐numpy实现

Comment (MARD1NO) [inline python/oneflow/test/graph/test_graph_adam_optim.py:44]: 我这里有个关于adam设置学习率的问题，我按照xiaoyu给的sgd示例，给adam增加generate_conf_for_graph这个函数。
然后我跑到`oneflow/user/kernels/model_update_kernel_util.h` AdamUpdateFunctor这里打log
    *v = next_v;
    *model =
        model_val - learning_rate * (next_m / (sqrt(next_v) + epsilon) + weight_decay * model_val);
    ...
    printf("==||| learning rate: %f \n", learning_rate); // Here!!!!!
然后我发现我在外面设置的是learning_rate = 1，这里打印结果是31.622810
设置learing_rate = 0.0001，打印结果是0.003162
就是这里的问题引发了计算对不齐numpy实现

File: python/oneflow/nn/optimizer/adam.py [hunk 5]
```diff
@@ -138,3 +139,38 @@ def step(self, closure: Callable = None):
...
+                False
+            )  # TODO(zzk): Check this option
+
+            for param in param_group.parameters:
+                vars_conf[param].l2 = l2
+                if not param.requires_grad:
+                    continue
+                optimizer_conf.add_variable_op_names(vars_conf[param].name)
```
Comment (strint) [inline python/oneflow/nn/optimizer/adam.py:176]: 这里之前写得有点绕，改成：
```
                if param.requires_grad:
                    optimizer_conf.add_variable_op_names(vars_conf[param].name)
```

Comment (MARD1NO) [reply python/oneflow/nn/optimizer/adam.py:176]: 这里我其实是抄你sgd的，我顺带也一起改了

File: python/oneflow/test/graph/test_graph_sgd_optim.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    train_by_numpy()
+
+    test_case.assertTrue(
+        np.allclose(np_res_list, of_res_list, rtol=0.001, atol=0.001)
+    )
+
+
+@flow.unittest.skip_unless_1n1d()
+class TestSGD(flow.unittest.TestCase):
+    def test_sgd1(test_case):
```
Comment (Ldpe2G) [inline python/oneflow/test/graph/test_graph_sgd_optim.py:96]: 改成和 test_adam 一样，多组参数组合

**Review by Ldpe2G — APPROVED** [review]

**Review by Ldpe2G — APPROVED** [review]
